### PR TITLE
Test `applyBlocks` with arbitrary sequences of transactions

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue number

ADP-1783

## Summary

This PR:
- introduces generators and shrinkers for arbitrary sequences of transactions.
- uses transaction sequences to test various properties of the [`applyBlocks`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L339) function.

## Reviewing

It's recommended to review this PR commit by commit. Earlier commits introduce abstract types that are specialised to more concrete types in later commits.

## New property tests

- [`prop_applyBlocks_filteredTxs_{all,none,some}Ours`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs#L2653)
    - tests that `applyBlocks` returns all transactions that are relevant to the wallet and no others.
    - tests that `applyBlocks` returns transactions in precisely the same order that they appear on the blockchain.
- [`prop_applyBlocks_lastUTxO_{all,none,some}Ours`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs#L2710)
    - tests that `applyBlocks` applies the UTxO transition rules correctly w.r.t.:
        - the subset of transactions that are relevant to the wallet.
        - the subset of the global UTxO set belonging to the wallet.

## New types

- [`StateDeltaSeq state delta`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/050b88ff3cbd8dd195e59208d103fe712695a607)
    - models a contiguous, abstract sequence of state transitions of the form:
      `state_0` → `delta_0_1` → `state_1` → `delta_1_2` → `state_2` → ... -> `state_n`
- [`TxSeq`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/7d780a6a1bdd8a8f87337cd63dac09d31c1728e5)
    - specialises a `StateDeltaSeq` so that `state ~ UTxO` and `delta ~ Tx`.
    - maintains an invariant that for every transition satisfies the UTxO transition rules.
    - provides support for partitioning sequences into sub-sequences. 
- [`BlockSeq`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/2ae205729e88d388fe5f322258dd7f0d1efc6ec8)
    - represents a sequence of blocks that starts at a given block height and slot number.
    - each block contains a sequence of transactions.

## Generating transaction sequences

See [`genTxSeq`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Types/TxSeq/Gen.hs#L91).

This PR generates transaction sequences according to the following simple algorithm:
1. Generate an initial sequence, consisting of a randomised initial UTxO and no transactions.
2. For `n` iterations:
   a. generate a random valid transaction `t` for the last UTxO in the sequence.
   b. apply transaction `t` to the last UTxO in the sequence, producing a updated sequence with an updated last UTxO.
   
## Shrinking transaction sequences

See [`shrinkTxSeq`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Types/TxSeq/Gen.hs#L120).

One of the biggest issues with testing properties involving sequences of transactions is that counterexamples can be very long and hard to interpret. A single transaction or UTxO set, when pretty printed, can easily fill half a screen. Long transaction sequences can easily fill many pages. Consequently, the pertinent part of a counterexample may be hidden within a huge sea of irrelevant details. In the event of a test failure, it's valuable to minimize the size of any counterexample so that just the pertinent information is presented.

When shrinking a `TxSeq`, it's important that all shrunken sequences maintain the UTxO transition rules. Therefore, individual transactions and UTxO sets cannot be shrunk in isolation: sequences must be shrunk as a whole.

There are many conceivable methods of shrinking a transaction sequence. This PR only provides support for some of these methods. Methods currently supported include:

 - shrinking to a prefix of the original sequence
 - shrinking to a suffix of the original sequence
 - removing block boundaries
 - removing individual assets
 - removing transactions that have no effect on the UTxO set
 - minimizing the set of assets that appear in UTxO sets and transactions
 - minimizing the set of transaction identifiers
 
The shrinker for `TxSeq` performs shrinking in phases, where each phase uses one of the above shrinking methods. Phases are ordered according to their aggressiveness (ability to shrink) and their efficiency (computation overhead): shrinking phases that are more aggressive and efficient are executed first.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs]: It's probably not useful to use `!` here. (These bangs were left over from a previous version of the branch.)

According to the documentation for `Seq`:
> Despite sequences being structurally strict from a semantic standpoint, they are in fact implemented using laziness internally.

**Comment by Unisay** [comment]
Suggested edit:
```diff
diff --git a/lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs b/lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs
index 682fc1934..b2f386d8b 100644
--- a/lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs
+++ b/lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs
@@ -161,7 +161,7 @@ shrinkTxWithoutId = genericRoundRobinShrink
     <:> liftShrink shrinkTxScriptValidity
     <:> Nil
 
-txWithoutIdToTx :: TxWithoutId -> Tx
+    = Tx {txId = mockHash tx, txCBOR = Nothing, ..}
 txWithoutIdToTx tx@TxWithoutId {..}
     = Tx {txId = mockHash tx, txCBOR = Nothing , ..}
 

```

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs
Comment (Unisay) [inline lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs]:
```suggestion
intersection (UTxO a) (UTxO b) = UTxO (Map.intersection a b)
```

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs
Comment (Unisay) [inline lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs]:
```suggestion
isProperSubsetOf (UTxO a) (UTxO b) = a `Map.isProperSubmapOf` b
```

File: lib/core/src/Data/Function/Utils.hs
Comment (Unisay) [inline lib/core/src/Data/Function/Utils.hs]:
```suggestion
isInjectiveOver f domain = Set.size domain == Set.size (Set.map f domain)
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]:
```suggestion
            it "prop_applyBlocks_filteredTxs_someOurs" $
                property prop_applyBlocks_filteredTxs_someOurs
            it "prop_applyBlocks_filteredTxs_allOurs" $
                property prop_applyBlocks_filteredTxs_allOurs
            it "prop_applyBlocks_filteredTxs_noneOurs" $
                property prop_applyBlocks_filteredTxs_noneOurs

        describe "lastUTxO" $ do
            it "prop_applyBlocks_lastUTxO_someOurs" $
                property prop_applyBlocks_lastUTxO_someOurs
            it "prop_applyBlocks_lastUTxO_allOurs" $
                property prop_applyBlocks_lastUTxO_allOurs
            it "prop_applyBlocks_lastUTxO_noneOurs" $
                property prop_applyBlocks_lastUTxO_noneOurs
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]:
```suggestion
    { initialBlockHeight :: Quantity "block" Word32
    , initialSlotNo :: SlotNo
    , shrinkableTxSeq :: ShrinkableTxSeq
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]:
```suggestion
            <$> ZipList (enumFrom $ initialBlockHeight blockSeq)
            <*> ZipList (enumFrom $ initialSlotNo blockSeq)
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: I suggest to consider using something like `error "headerHash is irrelevant"` to fail fast and precise if assumption doesn't hold.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: I am afraid that dummy values can mask a problem in case if some code suddenly relies on such a value.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: Actually, you already use it below, `shouldNotEvaluate "Block.headerHash"`

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

File: lib/test-utils/src/Test/QuickCheck/Extra.hs
Comment (Unisay) [inline lib/test-utils/src/Test/QuickCheck/Extra.hs:590]:
```suggestion
    lo = valueToCategorise `div` intervalSize * intervalSize
    hi = lo + pred intervalSize
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs]: I think I understand now what exactly you tried to achieve by putting `& property` in the end!
It makes a 1:1 alignment between the property function name and its test name. 
To me this looks like a job that is better done by machines, aka boilerplate. I'd either find a way to avoid having to repeat functions and their names or use the opportunity to provide a human-readable property description, e.g.:
```suggestion
            it "genTxSeq generates only valid transaction sequences" $
                prop_genTxSeq_isValid
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: @HeinrichApfelmus would be like
```suggestion
```
🤷🏼‍♂️

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (Unisay) [inline lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: Nice!

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs
Comment (Unisay) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs]: I don't understand how a  `Bounded a` could be greater than its max bound? 
```suggestion
    | a == maxBound = Nothing
```

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs
Comment (Unisay) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs]:
```suggestion
-- and intermediate UTxO sets, where each individual transaction 't_i_j' is
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (HeinrichApfelmus) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: Yup! I would be like that. I'm in favor of [PatternSynonyms](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/pattern_synonyms.html), but `ViewPatterns` just mixes up the order of arguments in a confusing way without providing much benefit in terms of abstraction capabilities.

File: lib/test-utils/src/Test/QuickCheck/Extra.hs
Comment (HeinrichApfelmus) [reply lib/test-utils/src/Test/QuickCheck/Extra.hs:590]: I would keep the parenthesis on the `lo` — the alternative parse

```
valueToCategorise `div` (intervalSize * intervalSize)
```

does not give the same result, and without parenthesis, one would have to look up the fixities of `div` and `*`.

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs]: The main justification for repeating the property name in the description is that when a property fails, it's very easy to search for the property function name and locate it in the codebase. Unfortunately, `hspec` doesn't print the property name by default when there's a failure.

I actually don't think that these description fields are the right place to describe what the property does. Very often, there's no easy way to write a concise description that fits on one line. Putting it on multiple lines often requires the use of awkward line continuation characters.

I do think that if we need to provide a natural language description of the property, then we'd derive more benefit from putting that description in a **_comment_** attached to the property definition itself. Comments have the advantage of allowing many lines, and code examples.

If you think there are any properties that are currently unclear and would benefit from comments, I'd be happy to add them. Let me know!

In an ideal world, I agree with you that we could just generate these `hspec` structures automatically, instead of hand-coding them. Though that's arguably outside of the scope of this PR!

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: @Unisay Thanks for your suggestion!

I'd prefer to keep the current style of this block (and others like it), as:
- it does not violate our [Coding Standards](https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md);
- we already have a precedent for using this style in many other spec modules;
- it's currently easy to see the correspondence between property name and the string identifier (see https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r974847186);
- it's currently easy to insert/delete modifiers post hoc, for example:
    ```patch
                it "prop_applyBlocks_lastUTxO_someOurs" $
                    prop_applyBlocks_lastUTxO_someOurs
    +                   & verboseShrinking
    +                   & withMaxSuccess 1_000
    +                   & within 1_000_000
                        & property
    ```

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: Thanks! I'm very grateful for the existence of the `quickcheck-classes` library.

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: @Unisay 

We do use `ViewPatterns` quite a bit throughout our codebase. (Around 15 modules already use it.)

Though I'd definitely be happy to adopt `PatternSynonyms` instead of `ViewPatterns` in future PRs!

Would you be happy with the following proposal:

Proposal:
1. Keep the current usage as it is (for this PR).
2. Create a follow-on PR that replaces all usages of `ViewPatterns` with `PatternSynonyms` (where feasible). That PR can also update other modules not included within the scope of this PR.

The proposal would unblock merging of this PR, and allow the replacement work to be prioritised independently. I'd be happy to do it as a gardening task. 👍🏻

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs]: > I don't understand how a Bounded a could be greater than its max bound?

In the general case, it's quite possible for someone to define `Bounded` and `Ord` instances for some type `t`, where there exists some value `a :: t` such that `a > maxBound`. It wouldn't be good practice to do this, but the compiler certainly wouldn't prevent it.

This successor function is only well-defined for values satisfying `< maxBound`. Therefore, it's not well-defined for values satisfying `>= maxBound`.

I'm not sure I see the advantage for limiting the boundary check to `==` here. An alternative to returning `Nothing` would be to raise a run-time error. But I cannot see a clear advantage, unless you can see one?

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs]: You clearly have the eyes of an eagle! 🦅

Thanks!

Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs]: Fixed in 4dafba07d0a37a397387cd363696372b90648bea.

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs]: @Unisay I agree that it's nice to avoid mixing infix- and prefix-style here.

I've converted these to use prefix style in 4dafba07d0a37a397387cd363696372b90648bea.

File: lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs]: I've converted this to use avoid infix application in 4dafba07d0a37a397387cd363696372b90648bea.

File: lib/core/src/Data/Function/Utils.hs
Comment (jonathanknowles) [reply lib/core/src/Data/Function/Utils.hs]: Fixed in 4dafba07d0a37a397387cd363696372b90648bea.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: @Unisay I think your suggested style is also fine, though the existing style doesn't violate the Coding Standards (to my knowledge). This style is also used in quite a few of our existing modules. Perhaps this discussion could be held outside the context of this PR?

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: I actually prefer the existing style here, as it emphasises the **_commonality_** between the two lines: in both cases we're enumerating from _attributes_ of `blockSeq`, and in both cases, we're enumerating from some `initial` value. If lenses were defined, we could have written:
```hs
            <$> ZipList (enumFrom $ blockSeq ^. #initialBlockHeight)
            <*> ZipList (enumFrom $ blockSeq ^. #initialSlotNo)
```

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: @Unisay I did actually use `error` when I first wrote this function. 😄 

However, then I discovered that `headerHash` is a strict field of `BlockHeader`, and `applyBlocks` pattern matches on values of `BlockHeader` in order to construct values of `ChainPoint`. 

So using `error` here is not an option, unfortunately, unless we change `applyBlocks` itself.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: > IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

Fair enough!

I had originally intended to extend this function with knowledge of how `applyBlocks` modifies fees, but eventually didn't get around to that. After giving the matter some thought, I'd probably be in favour of modifying `applyBlocks` so that it doesn't modify the `fee` field at all. (This behaviour seems completely unnecessary.)

I've inlined the definition and removed the function in bf0b7364f59fcb6c24dc4c9ae2b7348577c4346d.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: > Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

Given the current type of `applyBlocks`, I think it's useful to verify that the function doesn't modify transactions in any meaningful way.

If `applyBlocks` were modified to return a set of transaction identifiers (i.e., a subset of the blocks supplied), or if we had a type-level proof that the function was incapable of modifying transactions, then we could drop this comparison entirely.

File: lib/test-utils/src/Test/QuickCheck/Extra.hs
Comment (jonathanknowles) [reply lib/test-utils/src/Test/QuickCheck/Extra.hs:590]: @Unisay As @HeinrichApfelmus suggested, I included the parentheses to aid readability, to spare readers the pain of looking up the relative precedences of these operators. At least with parentheses, there can be no doubt in the mind of the reader as to the desired precedence.

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: Hi @Unisay

I do understand the point you make here, and I think it's a good one, but it raises a question: what values should we use here instead?

Internally, the `applyBlocks` function does pattern match on these values, so we can't conveniently use `error`.

File: lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs]: However, there is also this:
https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html

Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs]: However, it's currently difficult to experiment with changing this to [`Seq`](https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html) from [`strict-containers`](https://hackage.haskell.org/package/strict-containers-0.1), as the version of [`cardano-base`](https://github.com/input-output-hk/cardano-base) on which we currently depend also exports a library called [`strict-containers`](https://github.com/input-output-hk/cardano-base/tree/7e3ddba98a61900181fe63cdd4c9ed9708d1a6a7/strict-containers), and Cabal (currently) doesn't make it possible to depend upon two libraries with identical names.

This naming conflict was fixed in https://github.com/input-output-hk/cardano-base/pull/308, but it'll take a while for this change to become available in our environment.

So for now, I'll shelve the experiment of changing this to the strict version, and remove the bang operators.

Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs]: Fixed in 8d5bcaecd8dae708d27ddb7aa45b77e6d570c825.

File: lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs
Comment (HeinrichApfelmus) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: > The proposal would unblock merging of this PR, and allow the replacement work to be prioritised independently. I'd be happy to do it as a gardening task. 👍🏻

As long as `ViewPatterns` goes away, I'm fine with that.  😂

Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs]: @HeinrichApfelmus wrote:
> As long as ViewPatterns goes away, I'm fine with that. 😂 

Okay 😄. I've created a ticket for this: ADP-2230. (@HeinrichApfelmus, I've tentatively added you as co-pilot for that ticket. Does that work for you?)

My proposal is:
- For this PR, we keep the current usage as it is.
- After this PR is merged, I will create a **follow-on** PR that addresses ADP-2230 for the whole wallet. There are only 15 modules; it shouldn't take too long to replace all the usages with `PatternSynonyms`, or something else (on a case-by-case basis).

@Unisay, are you also okay with this plan?

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (Unisay) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: I see your point, usually I also prefer keeping similar parts on the lhs, and varying parts on rhs. 
In this particular case it leads to a mixture of `$` and `&` operators in the same expression and this makes it hard to parse in terms of associativity:
Example: `a $ b & c` - should it be associated to the left as `(a $ b) & c` or `a $ (b & c)` ?
Would you be willing to compromise on this:
```
            <$> ZipList (enumFrom (blockSeq & initialBlockHeight))
            <*> ZipList (enumFrom (blockSeq & initialSlotNo))
```
?

**Review by Unisay — APPROVED** [review]

File: lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs]: Hi @Unisay 

I've included your suggestion in a78e4215baffe74e50d199e8a64c9772d9bcebb4.

**Pull request merged**
